### DISCUSSION TITLE
Interpreting SymPy mechanics LagrangesMethod linearization output

### DISCUSSION THREAD

**Discussion by nikkujensen** [sympy/sympy#27409] [General]
Hello,

I have setup a problem using SymPy mechanics and have used the `LagrangesMethod ` to compute the equations of motion (EOMs). Now I want to linearize the EOMs, which I can successfully do using `LagrangesMethod.linearize()`. However, I don't quite understand the output.

Given the output of `LagrangesMethod.linearize()`, what is the 4th output? I notice in the example [here](https://docs.sympy.org/dev/tutorials/physics/mechanics/lin_pend_nonmin_example.html#lagrange-s-method) (when `A_and_B=True`) they store to a variable called `inp_vec`, but this is not recorded in the [linearize documentation](https://docs.sympy.org/dev/modules/physics/mechanics/api/linearize.html#sympy.physics.mechanics.linearize.Linearizer.linearize), though there is a brief mention in the [LagrangesMethod documentation](https://docs.sympy.org/dev/modules/physics/mechanics/api/kane_lagrange.html#sympy.physics.mechanics.lagrange.LagrangesMethod.linearize).

Is the 4th output `r` in the [LagrangesMethod documentation](https://docs.sympy.org/dev/modules/physics/mechanics/api/kane_lagrange.html#sympy.physics.mechanics.lagrange.LagrangesMethod.linearize)?

PS: Thank you to all who have developed the mechanics module! It has been very helpful for me work!

**Comment by anandkaranubc** [comment]
Hiii @nikkujensen,

I think you are right unless someone objects.

Based on the source code:

https://github.com/sympy/sympy/blob/b7e7de3b02c8d1edf683e12c61a69f57eebd4ce9/sympy/physics/mechanics/lagrange.py#L422-L423

```python
result = linearizer.linearize(**kwargs)
return result + (linearizer.r,)
```

- After getting the main results (usually \( M, A, B \)) from the `linearizer.linearize` method, the forcing terms \( r \) are added to the result as the fourth output.

- The method finally gives back a tuple with:
  - \( M, A, B, r \) (if `A_and_B=False`)
  - \( A, B, r \) (if `A_and_B=True`)


Here is where the checks for A and B are, if you are curious:

https://github.com/sympy/sympy/blob/b7e7de3b02c8d1edf683e12c61a69f57eebd4ce9/sympy/physics/mechanics/linearize.py#L422-L442

**Comment by moorepants** [comment]
The time varying variables in a system are either a state variable (coordinates or speeds) or a specified input (force, acceleration, voltage, etc). The prior commenter is correct that `r` (4th) output contains these exogenous inputs.
